### ISSUE TITLE
Using LSTM with AdaNet

### ISSUE THREAD

**Issue by mmuratarat** [tensorflow/adanet#58]
I can easily understand why DNN and CNN apply well to AdaNet since all subnetworks have the same architecture and hyperparameters. However, I am curious if it is possible to apply LSTM to AdaNet. I cannot fully comprehend how to incorporate `tf.nn.dtynamic_rnn()`, because when we have multiple LSTM cells it will be as the following:

outputs, states = tf.nn.dynamic_rnn([cell1, cell2,..., cell_n], inputs = X, swap_memory = False, time_major = False, dtype = tf.float32)

Can you provide any insight about that? Thank you so much!

**Comment by dataforcast** [comment]
Hello mmuratarat,
Hi,

I've worked on Adanet framework upon RNN type network (Simple RNN, LSTM and GRU).

In order to train mixture weights, Adanet may use logits or last layer from subnetwork candidates.
Then I force Adanet to use logits rather then last layer in creating Adanet Estimator with instructions :


```
from adanet.ensemble import MixtureWeightType
from adanet.ensemble.weighted import ComplexityRegularizedEnsembler

ensembler = ComplexityRegularizedEnsembler(mixture_weight_type=mixture_weight_type, adanet_lambda=1.e-3)
list_ensembler = [ensembler]

adanet_estimator = adanet.Estimator(
    subnetwork_generator=...,
    ensemblers   = [ensembler],    
    max_iteration_steps=500,
    subnetwork_generator = ...,
    head = ...,
    config= ...,    
    evaluator=adanet.Evaluator(
        input_fn=...,
        steps=None))
....
results, _ = tf.estimator.train_and_evaluate(adanet_estimator, train_spec, eval_spec)

```
Hope this will help

**Comment by cweill** [comment]
@mmuratarat: We've successfully train all kinds of RNNs with different cells like `lstm`, `cudnn_lstm`, and `gru`. Like @dataforcast mentions, you will need to create a custom `adanet.subnetwork.Generator` and `adanet.subnetwork.Builder` subclasses that use `tf.nn.dynamic_rnn()`. You can look at [SimpleDNN](https://github.com/tensorflow/adanet/blob/master/adanet/examples/simple_dnn.py) for inspiration.

**Comment by tx2010011751** [comment]
Hello! I'm a green hand in adanet. Would you please tell me if I have to build the LSTM model myself rather than using the LSTM API provided in kereas or tensorflow when using adanet? If so, would you please give an  example about using LSTM with adanet(the sample you use for test RNN). Thank you very much.

**Comment by cweill** [comment]
@tx2010011751 You should be able to use the [tf.contrib.estimator.RNNEstimator](https://www.tensorflow.org/api_docs/python/tf/contrib/estimator/RNNEstimator) with `adanet.AutoEnsembleEstimator` if you want to try ensembling RNN models. 

As a first step, try getting `tf.contrib.estimator.RNNEstimator` to train on its own, and next try it in adanet.

**Comment by le-dawg** [comment]
> @mmuratarat: We've successfully train all kinds of RNNs with different cells like `lstm`, `cudnn_lstm`, and `gru`. Like @dataforcast mentions, you will need to create a custom `adanet.subnetwork.Generator` and `adanet.subnetwork.Builder` subclasses that use `tf.nn.dynamic_rnn()`. You can look at [SimpleDNN](https://github.com/tensorflow/adanet/blob/master/adanet/examples/simple_dnn.py) for inspiration.

I need help in figuring out LSTM with AdaNet. I have followed the general structure of the SimpleDNN class and I got nowhere in _three_ days.

I simplified SimpleDNN as much as possible, ignoring the layer growth loop completely:
```
samples= list(features.values())[0]
# input_fn gives features,labels tuples, features.shape (non-tensor notation): 800 timesteps,7 features
# TensorShape([Dimension(None), Dimension(800), Dimension(7)])
    last_layer = tf.compat.v1.keras.layers.LSTM(10,
        #   return_sequences=True,
          activation=tf.nn.relu,
          kernel_initializer=tf.compat.v1.glorot_uniform_initializer(
              seed=self._seed))(images)

    last_layer = tf.compat.v1.layers.Dropout(
          rate=self._dropout, seed=self._seed)(last_layer)
    
    logits = tf.compat.v1.layers.Dense(
        units=logits_dimension,
        kernel_initializer=tf.compat.v1.glorot_uniform_initializer(
            seed=self._seed))(last_layer)

    # remove complexity for debugging
    # complexity = tf.sqrt(tf.cast(self._num_layers, dtype=tf.float32))
    complexity = tf.constant(1)
```
This is the only configuration that doesn't end up in [this conundrum](https://stackoverflow.com/questions/54732085/keras-tensor-reshaping-lstm-input-shape-error).
